File: packages/flutter/lib/src/logic/markdown.dart
/// Markdown 解析的 Dart 移植（对应 logic/markdown.ts）。
///
/// 与 Web 端同一套取舍：解析成 token 而不是 HTML 字符串，原始 HTML 当纯文本，
/// 地址过白名单。Flutter 这边尤其不该走「渲染 HTML」那条路——
/// 那要额外引一个 WebView 或 HTML 渲染库，把不可信内容交给它去解释。
import 'href.dart';

enum MdInlineType { text, code, strong, em, del, link, image }

class MdInline {
  const MdInline({
    required this.type,
    this.text = '',
    this.href = '',
    this.src = '',
    this.alt = '',
    this.children = const [],
  });

  final MdInlineType type;
  final String text;
  final String href;
  final String src;
  final String alt;
  final List<MdInline> children;
}

enum MdBlockType { heading, paragraph, code, quote, list, table, hr }

class MdBlock {
  const MdBlock({
    required this.type,
    this.level = 1,
    this.lang = '',
    this.text = '',
    this.open = false,
    this.ordered = false,
    this.start = 1,
    this.children = const [],
    this.blocks = const [],
    this.items = const [],
    this.head = const [],
    this.rows = const [],
    this.align = const [],
  });

  final MdBlockType type;
  final int level;
  final String lang;
  final String text;

  /// 流式过程中还没收到结尾围栏
  final bool open;
  final bool ordered;
  final int start;
  final List<MdInline> children;
  final List<MdBlock> blocks;
  final List<List<MdBlock>> items;
  final List<List<MdInline>> head;
  final List<List<List<MdInline>>> rows;

  /// 'left' / 'center' / 'right' / ''
  final List<String> align;
}

final _inline = RegExp(
  r'(`+)([\s\S]*?)\1'
  r'|!\[([^\]]*)\]\(((?:[^()]|\([^()]*\))*)\)'
  r'|\[([^\]]+)\]\(((?:[^()]|\([^()]*\))*)\)'
  r'|(\*\*|__)([\s\S]+?)\7'
  r'|(\*|_)([\s\S]+?)\9'
  r'|~~([\s\S]+?)~~',
);

/// 去掉地址后面的可选标题：`(/a "说明")`
///
/// 单引号用 \x27 写：仓库的 Dart 校验脚本按引号做字符串遮蔽来数括号，
/// 源码里直接出现的引号会让它把后面半行当成字符串，报一个与代码无关的错。
final _titleTail = RegExp(r'\s+(?:"|\x27|\()[\s\S]*$');

String _urlOf(String raw) => raw.replaceAll(_titleTail, '').trim();

/// 行内解析。代码优先于强调：粘一段代码进来不该被加粗一半。
List<MdInline> parseInline(String text) {
  final out = <MdInline>[];
  var rest = text;
  while (rest.isNotEmpty) {
    final hit = _inline.firstMatch(rest);
    if (hit == null) break;
    if (hit.start > 0) {
      out.add(MdInline(type: MdInlineType.text, text: rest.substring(0, hit.start)));
    }
    if (hit.group(2) != null) {
      out.add(MdInline(type: MdInlineType.code, text: hit.group(2)!.trim()));
    } else if (hit.group(4) != null) {
      final safe = safeHref(_urlOf(hit.group(4)!));
      final alt = hit.group(3) ?? '';
      if (safe != null) {
        out.add(MdInline(type: MdInlineType.image, src: safe, alt: alt));
      } else {
        out.add(MdInline(type: MdInlineType.text, text: alt.isEmpty ? hit.group(4)! : alt));
      }
    } else if (hit.group(6) != null) {
      final safe = safeHref(_urlOf(hit.group(6)!));
      final children = parseInline(hit.group(5)!);
      if (safe != null) {
        out.add(MdInline(type: MdInlineType.link, href: safe, children: children));
      } else {
        out.addAll(children);
      }
    } else if (hit.group(8) != null) {
      out.add(MdInline(type: MdInlineType.strong, children: parseInline(hit.group(8)!)));
    } else if (hit.group(10) != null) {
      out.add(MdInline(type: MdInlineType.em, children: parseInline(hit.group(10)!)));
    } else if (hit.group(11) != null) {
      out.add(MdInline(type: MdInlineType.del, children: parseInline(hit.group(11)!)));
    }
    rest = rest.substring(hit.end);
  }
  if (rest.isNotEmpty) out.add(MdInline(type: MdInlineType.text, text: rest));
  return out.isEmpty ? const [MdInline(type: MdInlineType.text)] : out;
}

final _hr = RegExp(r'^ {0,3}([-*_])(\s*\1){2,}\s*$');
final _listMark = RegExp(r'^(\s*)([-*+]|\d{1,9}[.)])\s+(.*)$');
final _heading = RegExp(r'^ {0,3}(#{1,6})\s+(.*)$');
final _fence = RegExp(r'^ {0,3}(`{3,}|~{3,})\s*([\w+-]*)\s*$');
final _quote = RegExp(r'^ {0,3}>\s?');

List<String> _tableRow(String line) => line
    .replaceFirst(RegExp(r'^\s*\|'), '')
    .replaceFirst(RegExp(r'\|\s*$'), '')
    .split('|')
    .map((cell) => cell.trim())
    .toList();

String _alignOf(String cell) {
  final left = cell.startsWith(':');
  final right = cell.endsWith(':');
  if (left && right) return 'center';
  if (right) return 'right';
  if (left) return 'left';
  return '';
}

/// 块级解析。停在代码块中间时按「还没收完」渲染（open: true），
/// 而不是把围栏当普通文字——否则用户会看到满屏反引号，写完的一瞬间又整段跳变。
List<MdBlock> parseMarkdown(String source) {
  final lines = source.replaceAll(RegExp(r'\r\n?'), '\n').split('\n');
  final blocks = <MdBlock>[];
  var i = 0;

  while (i < lines.length) {
    final line = lines[i];
    if (line.trim().isEmpty) {
      i += 1;
      continue;
    }

    final fence = _fence.firstMatch(line);
    if (fence != null) {
      final mark = fence.group(1)!;
      final lang = fence.group(2) ?? '';
      final body = <String>[];
      i += 1;
      var closed = false;
      final closing = RegExp('^ {0,3}\\${mark[0]}{${mark.length},}\\s*\$');
      while (i < lines.length) {
        if (closing.hasMatch(lines[i])) {
          closed = true;
          i += 1;
          break;
        }
        body.add(lines[i]);
        i += 1;
      }
      blocks.add(MdBlock(type: MdBlockType.code, lang: lang, text: body.join('\n'), open: !closed));
      continue;
    }

    if (_hr.hasMatch(line)) {
      blocks.add(const MdBlock(type: MdBlockType.hr));
      i += 1;
      continue;
    }

    final heading = _heading.firstMatch(line);
    if (heading != null) {
      blocks.add(MdBlock(
        type: MdBlockType.heading,
        level: heading.group(1)!.length,
        children: parseInline(heading.group(2)!.replaceFirst(RegExp(r'\s+#+\s*$'), '')),
      ));
      i += 1;
      continue;
    }

    if (_quote.hasMatch(line)) {
      final body = <String>[];
      while (i < lines.length && _quote.hasMatch(lines[i])) {
        body.add(lines[i].replaceFirst(_quote, ''));
        i += 1;
      }
      blocks.add(MdBlock(type: MdBlockType.quote, blocks: parseMarkdown(body.join('\n'))));
      continue;
    }

    if (line.contains('|') &&
        i + 1 < lines.length &&
        RegExp(r'^[\s|:-]+$').hasMatch(lines[i + 1]) &&
        lines[i + 1].contains('-')) {
      final head = _tableRow(line).map(parseInline).toList();
      final align = _tableRow(lines[i + 1]).map(_alignOf).toList();
      i += 2;
      final rows = <List<List<MdInline>>>[];
      while (i < lines.length && lines[i].contains('|') && lines[i].trim().isNotEmpty) {
        rows.add(_tableRow(lines[i]).map(parseInline).toList());
        i += 1;
      }
      blocks.add(MdBlock(type: MdBlockType.table, head: head, align: align, rows: rows));
      continue;
    }

    final mark = _listMark.firstMatch(line);
    if (mark != null) {
      final ordered = RegExp(r'\d').hasMatch(mark.group(2)!);
      final start = ordered ? int.parse(RegExp(r'\d+').firstMatch(mark.group(2)!)!.group(0)!) : 1;
      final baseIndent = mark.group(1)!.length;
      final items = <List<MdBlock>>[];
      while (i < lines.length) {
        final current = _listMark.firstMatch(lines[i]);
        if (current == null || current.group(1)!.length < baseIndent) break;
        if (RegExp(r'\d').hasMatch(current.group(2)!) != ordered &&
            current.group(1)!.length == baseIndent) {
          break;
        }
        if (current.group(1)!.length > baseIndent) {
          final nested = <String>[];
          while (i < lines.length) {
            final deeper = _listMark.firstMatch(lines[i]);
            if (deeper == null || deeper.group(1)!.length <= baseIndent) break;
            nested.add(lines[i].substring(baseIndent + 2));
            i += 1;
          }
          if (items.isNotEmpty) items.last.addAll(parseMarkdown(nested.join('\n')));
          continue;
        }
        items.add(parseMarkdown(current.group(3)!));
        i += 1;
      }
      blocks.add(MdBlock(type: MdBlockType.list, ordered: ordered, start: start, items: items));
      continue;
    }

    final paragraph = <String>[];
    while (i < lines.length &&
        lines[i].trim().isNotEmpty &&
        _listMark.firstMatch(lines[i]) == null &&
        !_hr.hasMatch(lines[i]) &&
        !_heading.hasMatch(lines[i]) &&
        !RegExp(r'^ {0,3}(`{3,}|~{3,})').hasMatch(lines[i]) &&
        !_quote.hasMatch(lines[i])) {
      paragraph.add(lines[i]);
      i += 1;
    }
    if (paragraph.isNotEmpty) {
      blocks.add(MdBlock(type: MdBlockType.paragraph, children: parseInline(paragraph.join('\n'))));
    } else {
      i += 1;
    }
  }

  return blocks;
}

String _inlineText(List<MdInline> nodes) => nodes.map((n) {
      if (n.type == MdInlineType.text || n.type == MdInlineType.code) return n.text;
      if (n.type == MdInlineType.image) return n.alt;
      return _inlineText(n.children);
    }).join();

/// 取纯文本，用于摘要、搜索与复制
String markdownText(List<MdBlock> blocks) => blocks
    .map((block) {
      switch (block.type) {
        case MdBlockType.heading:
        case MdBlockType.paragraph:
          return _inlineText(block.children);
        case MdBlockType.code:
          return block.text;
        case MdBlockType.quote:
          return markdownText(block.blocks);
        case MdBlockType.list:
          return block.items.map(markdownText).join('\n');
        case MdBlockType.table:
          return [block.head, ...block.rows]
              .map((row) => row.map(_inlineText).join('\t'))
              .join('\n');
        case MdBlockType.hr:
          return '';
      }
    })
    .where((text) => text.isNotEmpty)
    .join('\n\n');

File: packages/common/src/logic/markdown.ts
/**
 * Markdown → 结构化 token，**不产出 HTML 字符串**。
 *
 * 这一层的设计只有一个出发点：内容来自模型与后端，也就是不可信的地方。
 * 主流做法是「渲染成 HTML 再清洗」，那要求清洗器穷举所有变形，
 * 漏一种就等于给了对方一个 XSS；而且 HTML 字符串没法在小程序与 Flutter 上渲染，
 * 各端只能各写一套解析。
 *
 * 所以这里解析成 token 树，由各端用自己的原生元素渲染：
 * 原始 HTML 一律当纯文本，地址一律过 safeHref——这两条不是可配置项。
 *
 * 支持的语法按「模型真的会输出什么」取舍：标题、段落、列表（含嵌套一层）、
 * 有序列表、代码块、行内代码、引用、表格、分割线、粗体、斜体、删除线、
 * 链接、图片。不支持脚注、定义列表、HTML 内嵌——它们在对话里几乎不出现，
 * 而每多支持一种就多一处要防的地方。
 */
import { safeHref } from './href'

export type MdInline =
  | { type: 'text'; text: string }
  | { type: 'code'; text: string }
  | { type: 'strong'; children: MdInline[] }
  | { type: 'em'; children: MdInline[] }
  | { type: 'del'; children: MdInline[] }
  | { type: 'link'; href: string; children: MdInline[] }
  | { type: 'image'; src: string; alt: string }

export type MdBlock =
  | { type: 'heading'; level: 1 | 2 | 3 | 4 | 5 | 6; children: MdInline[] }
  | { type: 'paragraph'; children: MdInline[] }
  | { type: 'code'; lang: string; text: string; /** 流式过程中还没收到结尾围栏 */ open: boolean }
  | { type: 'quote'; children: MdBlock[] }
  | { type: 'list'; ordered: boolean; start: number; items: MdBlock[][] }
  | { type: 'table'; head: MdInline[][]; rows: MdInline[][][]; align: ('left' | 'center' | 'right' | null)[] }
  | { type: 'hr' }

/* ───────────────────────── 行内 ───────────────────────── */

/*
 * 地址部分要一路吃到配对的右括号，而不是「吃到第一个右括号为止」。
 *
 * 后者有两个真实后果：`(javascript:alert(1))` 会在第一个 `)` 处截断，
 * 剩下一个孤零零的 `)` 落到正文里；更糟的是 `(java\tscript:alert(1))`
 * ——按「不含空白」截取会得到地址 `java`，那是个合法的相对路径，
 * 于是一个恶意地址被解析成了一个看似正常的链接。所以整段原样交给 safeHref，
 * 由它统一拒绝，而不是在这里先切一刀。
 */
const URL_PART = '((?:[^()]|\\([^()]*\\))*)'
const INLINE = new RegExp(
  '(`+)([\\s\\S]*?)\\1' +
  `|!\\[([^\\]]*)\\]\\(${URL_PART}\\)` +
  `|\\[([^\\]]+)\\]\\(${URL_PART}\\)` +
  '|(\\*\\*|__)([\\s\\S]+?)\\7' +
  '|(\\*|_)([\\s\\S]+?)\\9' +
  '|~~([\\s\\S]+?)~~'
)

/** 去掉地址后面的可选标题：`(/a "说明")` */
const urlOf = (raw: string) => raw.replace(/\s+["'(][\s\S]*$/, '').trim()

/**
 * 行内解析。
 *
 * 代码优先：`` `**不是粗体**` `` 里的星号必须原样显示，否则粘一段代码进来，
 * 界面上会莫名其妙地加粗一半。
 */
export function parseInline(text: string): MdInline[] {
  const out: MdInline[] = []
  let rest = text
  while (rest) {
    const hit = INLINE.exec(rest)
    if (!hit || hit.index === undefined) break
    if (hit.index > 0) out.push({ type: 'text', text: rest.slice(0, hit.index) })
    const [matched, , codeText, alt, src, linkText, href, , strongText, , emText, delText] = hit

    if (codeText !== undefined) {
      out.push({ type: 'code', text: codeText.trim() })
    } else if (src !== undefined) {
      // 地址不安全时退化成替代文本：渲染一个点了会出事的图不如不渲染
      const safe = safeHref(urlOf(src))
      if (safe) out.push({ type: 'image', src: safe, alt: alt ?? '' })
      else out.push({ type: 'text', text: alt || src })
    } else if (href !== undefined) {
      const safe = safeHref(urlOf(href))
      if (safe) out.push({ type: 'link', href: safe, children: parseInline(linkText) })
      else out.push(...parseInline(linkText))
    } else if (strongText !== undefined) {
      out.push({ type: 'strong', children: parseInline(strongText) })
    } else if (emText !== undefined) {
      out.push({ type: 'em', children: parseInline(emText) })
    } else if (delText !== undefined) {
      out.push({ type: 'del', children: parseInline(delText) })
    }
    rest = rest.slice(hit.index + matched.length)
  }
  if (rest) out.push({ type: 'text', text: rest })
  return out.length ? out : [{ type: 'text', text: '' }]
}

/* ───────────────────────── 块级 ───────────────────────── */

const isHr = (line: string) => /^ {0,3}([-*_])(\s*\1){2,}\s*$/.test(line)
const listMark = (line: string) => /^(\s*)([-*+]|\d{1,9}[.)])\s+(.*)$/.exec(line)

function parseTableRow(line: string): string[] {
  return line
    .replace(/^\s*\|/, '')
    .replace(/\|\s*$/, '')
    .split('|')
    .map((cell) => cell.trim())
}

const alignOf = (cell: string): 'left' | 'center' | 'right' | null => {
  const left = cell.startsWith(':')
  const right = cell.endsWith(':')
  if (left && right) return 'center'
  if (right) return 'right'
  if (left) return 'left'
  return null
}

/**
 * 块级解析。
 *
 * 流式是常态而不是例外：模型一边写一边渲染，随时可能停在代码块中间。
 * 停在中间时按「还没收完的代码块」渲染（open: true），而不是把围栏当普通文字，
 * 否则用户会看到满屏的反引号，写完的一瞬间又整段跳变。
 */
export function parseMarkdown(source: string): MdBlock[] {
  const lines = source.replace(/\r\n?/g, '\n').split('\n')
  const blocks: MdBlock[] = []
  let i = 0

  while (i < lines.length) {
    const line = lines[i]

    if (!line.trim()) { i += 1; continue }

    // 代码围栏
    const fence = /^ {0,3}(`{3,}|~{3,})\s*([\w+-]*)\s*$/.exec(line)
    if (fence) {
      const [, mark, lang] = fence
      const body: string[] = []
      i += 1
      let closed = false
      while (i < lines.length) {
        if (new RegExp(`^ {0,3}${mark[0]}{${mark.length},}\\s*$`).test(lines[i])) { closed = true; i += 1; break }
        body.push(lines[i])
        i += 1
      }
      blocks.push({ type: 'code', lang: lang || '', text: body.join('\n'), open: !closed })
      continue
    }

    if (isHr(line)) { blocks.push({ type: 'hr' }); i += 1; continue }

    const heading = /^ {0,3}(#{1,6})\s+(.*)$/.exec(line)
    if (heading) {
      blocks.push({
        type: 'heading',
        level: heading[1].length as 1 | 2 | 3 | 4 | 5 | 6,
        children: parseInline(heading[2].replace(/\s+#+\s*$/, ''))
      })
      i += 1
      continue
    }

    if (/^ {0,3}>\s?/.test(line)) {
      const body: string[] = []
      while (i < lines.length && (/^ {0,3}>\s?/.test(lines[i]) || lines[i].trim())) {
        if (!/^ {0,3}>\s?/.test(lines[i])) break
        body.push(lines[i].replace(/^ {0,3}>\s?/, ''))
        i += 1
      }
      blocks.push({ type: 'quote', children: parseMarkdown(body.join('\n')) })
      continue
    }

    // 表格：第二行必须是分隔行，否则那只是一段带竖线的普通文字
    if (line.includes('|') && i + 1 < lines.length && /^[\s|:-]+$/.test(lines[i + 1]) && lines[i + 1].includes('-')) {
      const head = parseTableRow(line).map(parseInline)
      const align = parseTableRow(lines[i + 1]).map(alignOf)
      i += 2
      const rows: MdInline[][][] = []
      while (i < lines.length && lines[i].includes('|') && lines[i].trim()) {
        rows.push(parseTableRow(lines[i]).map(parseInline))
        i += 1
      }
      blocks.push({ type: 'table', head, align, rows })
      continue
    }

    const mark = listMark(line)
    if (mark) {
      const ordered = /\d/.test(mark[2])
      const start = ordered ? Number.parseInt(mark[2], 10) : 1
      const items: MdBlock[][] = []
      const baseIndent = mark[1].length
      while (i < lines.length) {
        const current = listMark(lines[i])
        if (!current || current[1].length < baseIndent) break
        if (/\d/.test(current[2]) !== ordered && current[1].length === baseIndent) break
        if (current[1].length > baseIndent) {
          // 缩进的行属于上一项：收集起来递归解析，嵌套列表就是这么来的
          const nested: string[] = []
          while (i < lines.length) {
            const deeper = listMark(lines[i])
            if (!deeper || deeper[1].length <= baseIndent) break
            nested.push(lines[i].slice(baseIndent + 2))
            i += 1
          }
          const last = items[items.length - 1]
          if (last) last.push(...parseMarkdown(nested.join('\n')))
          continue
        }
        items.push(parseMarkdown(current[3]))
        i += 1
      }
      blocks.push({ type: 'list', ordered, start, items })
      continue
    }

    // 段落：连续的非空行合成一段
    const paragraph: string[] = []
    while (i < lines.length && lines[i].trim() && !listMark(lines[i]) && !isHr(lines[i])
      && !/^ {0,3}(#{1,6})\s/.test(lines[i]) && !/^ {0,3}(`{3,}|~{3,})/.test(lines[i])
      && !/^ {0,3}>\s?/.test(lines[i])) {
      paragraph.push(lines[i])
      i += 1
    }
    if (paragraph.length) blocks.push({ type: 'paragraph', children: parseInline(paragraph.join('\n')) })
    else i += 1
  }

  return blocks
}

/** 取纯文本，用于摘要、搜索与复制 */
export function markdownText(blocks: MdBlock[]): string {
  const inline = (nodes: MdInline[]): string =>
    nodes.map((n) => {
      if (n.type === 'text' || n.type === 'code') return n.text
      if (n.type === 'image') return n.alt
      return inline(n.children)
    }).join('')

  return blocks.map((block) => {
    switch (block.type) {
      case 'heading':
      case 'paragraph':
        return inline(block.children)
      case 'code':
        return block.text
      case 'quote':
        return markdownText(block.children)
      case 'list':
        return block.items.map((item) => markdownText(item)).join('\n')
      case 'table':
        return [block.head, ...block.rows].map((row) => row.map(inline).join('\t')).join('\n')
      default:
        return ''
    }
  }).filter(Boolean).join('\n\n')
}
